// Protocol messages for compiler IRs.
//
// Copyright 2019-2020 the ProGraML authors.
//
// Contact Chris Cummins <chrisc.101@gmail.com>.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package programl;

option cc_enable_arenas = true;
option go_package = "programlpb";
option java_multiple_files = true;
option java_outer_classname = "UtilProto";
option java_package = "com.programl";

import "programl/proto/program_graph.proto";
import "programl/third_party/tensorflow/features.proto";

// A list of program graphs.
message ProgramGraphList {
  // A `<key, value>` mapping of features for this ProgramGraph list.
  Features context = 1;
  // A list of ProgramGraph messages.
  repeated ProgramGraph graph = 2;
}

// Options used to generate a program graph.
message ProgramGraphOptions {
  // If set, the program graph builder will reject graphs where:
  //  1. A module contains no nodes.
  //  2. A function contains no nodes.
  //  3. A node is unnconnected.
  bool strict = 3;
  // Generate only nodes for instructions.
  bool instructions_only = 1;
  // Omit return call edges from call sites.
  bool ignore_call_returns = 2;
  // The optimization level when generating an IR from a source file.
  int32 opt_level = 4;
  // The path of an IR to read.
  string ir_path = 10;
}

// Features describing a program.
message ProgramGraphFeatures {
  // A list of features corresponding a ProgramGraph's list of Node messages.
  FeatureLists node_features = 1;
  // A list of features corresponding a ProgramGraph's list of Edge messages.
  FeatureLists edge_features = 2;
  // A list of features corresponding a ProgramGraph's list of Function messages.
  FeatureLists function_features = 3;
  // A list of features corresponding a ProgramGraph's list of Module messages.
  FeatureLists module_features = 4;
  // A set of graph-level features.
  Features features = 5;
}

// A list of program graphs.
message ProgramGraphFeaturesList {
  // A `<key, value>` mapping of features for this ProgramGraphFeatures list.
  Features context = 1;
  // A list of ProgramGraphFeatures messages.
  repeated ProgramGraphFeatures graph = 2;
}

// A compiler intermediate representation.
message Ir {
  // The type of IR.
  enum Type {
    UNKNOWN = 0;
    LLVM = 1;
    XLA_HLO = 2;
  }
  // The type of IR.
  Type type = 1;
  // The compiler version, as a single integer. Major + minor versions
  // should be converted to this single number, e.g. 6.0.0 -> 600.
  int64 compiler_version = 2;
  // The command that was used to produce this IR.
  string cmd = 3;
  // The text of the IR.
  string text = 4;
}

// A list of compiler IRs.
message IrList {
  // A list of Ir messages.
  repeated Ir ir = 1;
}

// A source file.
message SourceFile {
  // The source programming language.
  enum Language {
    UNKNOWN = 0;
    C = 1;
    CXX = 2;
    OPENCL = 3;
    SWIFT = 4;
    HASKELL = 5;
    FORTRAN = 6;
  }
  // The source programming language.
  Language language = 2;
  // The relative path of the file.
  string relpath = 1;
  // The text of the file.
  string text = 3;
}

// A repository of source files.
message Repo {
  // The URL of the repository.
  string url = 1;
  // The sha1 of the repository's HEAD commit.
  string sha1 = 2;
  // The timestamp that this repository was created.
  int64 created_ms_timestamp = 3;
}

// A node map is used to translate node indices between ProgramGraph instances.
message NodeIndexList {
  // The keys are node indices in the old graph representation, the
  // values are node inidices in the new graph representation.
  repeated int32 node = 1;
}
